File: incre-tests/synduce/tree/sorted_2.f
Inductive Tree = leaf Int | node {Int, Tree, Tree};

spec = fix (
  \f: Int -> Tree -> Bool. \y: Int. \t: Tree.
  match t with
    leaf a -> > y a
  | node {a, l, r} ->
    let r1 = (f a l) in
    let r2 = (f a r) in
    and r1 (and r2 (> y a))
  end
);

repr = fix (
  \f: {Int, Tree} -> Compress {Int, Tree}. \t: {Int, Tree}.
  match t with
    {pre, leaf a} -> {pre, leaf a}
  | {pre, node {a, l, r}} ->
    let lres = f {a, l} in
      let rres = f {a, r} in
        {pre, node {a, lres.2, rres.2}}
  end
);

main = \key: Int. \t: Tree.
  let res = repr {key, t} in
    spec key res.2;

File: ml-benchmark/synduce/tree/sorted_2.ml
type tree =
  | Leaf of int
  | Node of int * tree * tree

let rec spec y t =
  match t with
  | Leaf a -> y > a
  | Node (a, l, r) ->
      let r1 = spec a l in
      let r2 = spec a r in
      r1 && (r2 && (y > a))

let rec repr t =
  match t with
  | (pre, tt) ->
      match tt with
      | Leaf a -> (pre, Leaf a)
      | Node (a, l, r) ->
          let lres = repr (a, l) in
          let rres = repr (a, r) in
          let ltree =
            match lres with
            | (_, lt) -> lt
          in
          let rtree =
            match rres with
            | (_, rt) -> rt
          in
          (pre, Node (a, ltree, rtree))

let program key t =
  let res = repr (key, t) in
  match res with
  | (_, t2) -> spec key t2